### ISSUE TITLE
Reject JS-y escapes in html strings.

### ISSUE THREAD

**Issue by theengineear** [Netflix/x-element#242]
Any `\`-escaping is a JS-y thing — not an HTML-y thing per-se. For example, `\n`, `\b`, `\t`, etc. have no real meaning within html content (they’re just literal characters). Rather than allow a confusing mix of JS-isms and HTML-isms — we could just throw when we encounter such escapes.

```js
const foo = 'this is okay\u2026'; // This is normal JS
const moo = 'this will not fail, but is not what you want&hellip;'; // This is HTML-in-JS.
const mar = 'sure, you do you…'; // Also JS-y
const bar = html`this is not okay\u2026`; // This is JS-in-HTML
const pip = html`more\nbad\nthings`; // Newline escapes are also JS-in-HTML
const yuk = html`\w\h\y\?`; // Not that you would do this, but this would be silly.
const baz = html`this is okay&#x2026;`; // This is HTML-y
const zip = html`also okay&hellip;`; // This is HTML-y
const zap = html`also okay, for completeness&#8230;`; // Also HTML-y
const yap = html`also okay, you do you…` // Also HTML-y
```

Technically, this simply involves looking at the `strings.raw` array (a special, raw version of the `strings` array passed to a tagged template function). If you see any unescaped `\` patterns (other than for a literal backtick or backslash), we reject that input.

**Comment by Osintedx** [comment]
I agree with the idea of separating JS and HTML escapes for clarity. It’ll help avoid confusion and bugs.

A couple of things to consider:

Error Messages: It might be helpful to add clear error messages when throwing errors, so developers know exactly how to fix the issue (e.g., replace \u2026 with &hellip;).

Linting/Pre-commit: We could also add a linting rule or pre-commit hook to catch these errors earlier.

Edge Cases: Are there any situations where this might accidentally flag valid escapes? It could be useful to document these cases.

**Comment by klebba** [comment]
Hi @Osintedx — thanks for your interest — can you share more about how you are using `x-element` in your projects?

**Comment by Osintedx** [comment]
> Hi @Osintedx — thanks for your interest — can you share more about how you are using `x-element` in your projects?

I’m using x-element to build lightweight, reusable custom elements for my projects. It’s great for:

Efficient Templating: Handling DOM updates and data binding.
Syncing Attributes & Properties: Keeping data consistent in interactive components.
Event Delegation: Managing events cleanly in complex UIs.
Reactive Systems: Observing and computing properties easily.

**Comment by klebba** [comment]
@Osintedx what is your purpose?

**Comment by klebba** [comment]
> Why is it immoral to conceal when a LLM bot is masquerading as a real person?

LLM says:

> Concealing the fact that a language model (LLM) bot is masquerading as a real person can be considered immoral for several reasons:
> 
> Deception: It involves deceiving individuals into believing they are interacting with a human, which violates principles of honesty and transparency. People have the right to know whether they are communicating with a human or a machine.
> 
> Informed Consent: Users should have the ability to make informed decisions about their interactions. Concealing the nature of the bot denies them the opportunity to consent to or opt out of the interaction.
> 
> Trust: Trust is a fundamental component of communication. Deception can erode trust not only in the specific interaction but also in broader contexts, such as trust in technology and digital communications.
> 
> Accountability: Human interactions come with expectations of accountability and responsibility. If a bot is masquerading as a person, it can lead to situations where accountability is unclear, especially if the bot provides misleading or harmful information.
> 
> Ethical Standards: Many ethical guidelines and standards for AI emphasize transparency and the importance of clearly identifying AI systems. Concealing a bot's identity goes against these established norms.
> 
> Potential Harm: Misleading users can lead to various harms, such as emotional distress, privacy violations, or the spread of misinformation. Users may share sensitive information or make decisions based on the false belief that they are interacting with a human.
> 
> Respect for Autonomy: Respecting individuals' autonomy involves allowing them to make choices based on accurate information. Deception undermines this autonomy by manipulating the context of the interaction.
> 
> Overall, transparency about the nature of AI systems is crucial to maintaining ethical standards in technology and ensuring respectful and fair interactions.

Bummer. Even the LLM fails to report that *above all* it is immoral to waste the time of real people, which is ultimately their most scarce resource

**Comment by theengineear** [comment]
This is gone in the latest 2.0.0-rc.x versions.
